File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-6.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        queue<tuple<TreeNode*, int>> t_queue;
        vector<vector<int>> res;

        if (root) {
            t_queue.push({root, 0});
        }

        while (t_queue.size() > 0) {
            auto [node, level] = t_queue.front();
            t_queue.pop();

            if (res.size() <= level) {
                res.push_back(vector<int>());
            }

            res[level].push_back(node->val);

            if (node->left) {
                t_queue.push({node->left, level + 1});
            }

            if (node->right) {
                t_queue.push({node->right, level + 1});
            }
        }

        return res;
    }
};

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-6.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.path_sum = root.val

        def dfs(root):
            if not root:
                return 0
            
            left_path = dfs(root.left)
            right_path = dfs(root.right)
            # small_path = left_path + right_path + root.val
            # max_val = max(root.val, root.val + left_path, root.val + right_path)
            # max_val = small_path if small_path > max_val else max_val
            left_max = max(left_path, 0)
            right_max = max(right_path, 0)
            self.path_sum = max(self.path_sum, root.val + left_max + right_max)

            return root.val + max(left_max, right_max)
        
        dfs(root)

        return self.path_sum
        